### ISSUE TITLE
EMSDK support for chrome extension manifest version v3

### ISSUE THREAD

**Issue by aykejriw** [emscripten-core/emscripten#16969]
We have been using emsdk to build wasm/js files to be used in a chrome extension manifest version v2.

As per guidance from Google to migrate from v2 to v3 [https://developer.chrome.com/docs/extensions/whatsnew/](url), our wasm/js files have started giving errors like - 
1. wasm streaming compile failed: TypeError: Failed to fetch falling back to ArrayBuffer instantiation
2. XMLHttpRequest is not defined
    at Object.readBinary
    at getBinary

We have been compiling using these emsdk flags - WASM = 1, MODULARIZE =1, EXPORT_ALL = 1, EXPORT_NAME = 'library', DYNAMIC_EXECUTION=0

Is there any latest change in emsdk supporting manifest version v3 for chrome extensions? 
If I am not wrong, deadline to migrate to v3 is Jan'23.

Please let us know if we should add some flags while compiling.

**Comment by kripken** [comment]
> Is there any latest change in emsdk supporting manifest version v3 for chrome extensions?

I'm not aware of any work specifically on manifest v3.

I know almost nothing about manifest v3, but I'd guess we need to debug at the JS level why the fetch fails? Perhaps there is something in the devtools log?

**Comment by lnxjnky** [comment]
@kripken It seems with manifest v3 in extension instead of constantly running background script, service workers are introduced. Also, the use of XMLHttpRequest and creation of new Worker is not allowed inside the service worker. 

Object for XHR and Worker are undefined when created in context of service worker.

The wasm/js code which gets generated using emsdk uses both XHR and worker (for pthreads). This now creates problem while running wasm/js as part of service worker.

Do you have any guidance how we can resolve this?

**Comment by kripken** [comment]
If service workers do not allow creation of Web Workers then pthreads can't be used there. I don't think there's a workaround for that. This would be a limitation of service workers (and/or manifest v3, if it forces you to use service workers).

If XHR is a problem, perhaps that can be worked around. Does `fetch` work in those places?

**Comment by lnxjnky** [comment]
Yes fetch works fine for asynchronous calls, but I see that there are also sync calls being made from emscripten generated wasm/js files via XHR.

Pasting the snippet from emscripten generated file:

```
read_ = url => {
            var xhr = new XMLHttpRequest;
            xhr.open("GET", url, false);
            xhr.send(null);
            return xhr.responseText
          };
          if (ENVIRONMENT_IS_WORKER) {
            readBinary = url => {
              var xhr = new XMLHttpRequest;
              xhr.open("GET", url, false);
              xhr.responseType = "arraybuffer";
              xhr.send(null);
              return new Uint8Array(xhr.response)
            }
          }
```

**Comment by kripken** [comment]
@sbc100 Yes, makes sense to do that.

But fetch can't be sync, though, so it wouldn't be enough to fix @lnxjnky 's problem. We'd need to also avoid sync. First thing would be to find out why readBinary is called - that must mean the async path to load the wasm failed.

**Comment by punithbm** [comment]
@sbc100  @kripken  I am also facing a similar issue when loading wasm as mentioned here - https://github.com/trustwallet/wallet-core/issues/2429 

Is there any workaround or any suggestion which could help me to run this script with manifest version 3? cause chrome has stopped accepting the new apps with manifest version 2 and the only option we have right now is version 3. Please let me know.

**Comment by kripken** [comment]
@punithbm I'm afraid none of us here are experts on manifest v3, and reading your linked issue, the problem is "refused to execute inline script [..] script-src 'self'". I'm not familiar with that error (is "self" not allowed?), but as a first step I suggest finding out the failing line of code.

**Comment by emaxx-google** [comment]
https://issues.chromium.org/issues/40772041#comment24 suggests a workaround for the Web Worker instantiation from manifestv3 extensions: use an Offscreen Document.

Does anyone have ideas on how this trick could be injected into Emscripten's Web Worker management code?
